frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2244,6 +2262,306 @@ void resolveUsesAndImportsInScope(Context* context, const Scope* scope) {
+          nConstRef++;
+          break;
+        case Function::REF:
+          nRef++;
+          break;
+        case Function::PARAM:
+        case Function::TYPE:
+          nOther++;
+          break;
+      }
+    }
+  }
+}
+
+
+static const bool&
+emitMultipleDefinedSymbolErrorsQuery(Context* context, const Scope* scope) {
+  QUERY_BEGIN(emitMultipleDefinedSymbolErrorsQuery, context, scope);
+
+  bool result = false;
+
+  std::set<UniqueString> namesDefined;
+  std::set<UniqueString> namesDefinedMultiply;
+  ScopeSet checkedScopes;
+
+  collectAllNames(context, scope, /* skip private */ false,
+                  namesDefined, namesDefinedMultiply, checkedScopes);
+
+  // Now, consider names in namesDefinedMultiply. If there are any
+  // that are not only parenful functions, issue an error.
+  LookupConfig config = LOOKUP_DECLS |
+                        LOOKUP_IMPORT_AND_USE |
+                        LOOKUP_SKIP_SHADOW_SCOPES;
+  for (auto name : namesDefinedMultiply) {
+    std::vector<BorrowedIdsWithName> v =
+      lookupNameInScope(context, scope, { }, name, config);
+    int nParenfulMethods = 0;
+    int nParenfulNonMethodFunctions = 0;
+    int nParenlessMethods = 0;
+    int nParenlessNonMethodFunctions = 0;
+    int nFields = 0;
+    int nOther = 0;
+    countSymbols(context, v,
+                 nParenfulMethods, nParenfulNonMethodFunctions,
+                 nParenlessMethods, nParenlessNonMethodFunctions,
+                 nFields, nOther);
+    int nNonFunctions = nFields + nOther;
+    bool error = false;
+    if (nParenfulNonMethodFunctions > 0 &&
+        nParenlessNonMethodFunctions + nNonFunctions > 0) {
+      // mix of parenful functions and parenless functions / non-function
+      error = true;
+    } else if (nParenlessNonMethodFunctions > 0 && nNonFunctions  > 0) {
+      // mix of parenless functions and non-functions
+      error = true;
+    } else if (nNonFunctions > 1) {
+      // multiple non-functions with the same name
+      error = true;
+    } else if (nParenlessNonMethodFunctions > 1) {
+      // multiple parenless non-method functions with the same name
+      // check: is it return intent overloading?
+      int nValue = 0;
+      int nConstRef = 0;
+      int nRef = 0;
+      int nOther = 0;
+      countReturnIntents(context, v, nValue, nConstRef, nRef, nOther);
+      if (nValue <= 1 && nConstRef <= 1 && nRef <= 1 && nOther == 0) {
+        // it is a valid return intent overload
+      } else {
+        error = true;
+      }
+    }
+    if (error) {
+      // emit a multiply-defined symbol error
+      std::vector<ResultVisibilityTrace> traceResult;
+      v = lookupNameInScopeTracing(context, scope, { }, name, config,
+                                   traceResult);
+
+      CHPL_REPORT(context, Redefinition, scope->id(), name, v, traceResult);
+
+      result = true;
+    }
+  }
+
+  return QUERY_END(result);
+}
+
+void emitMultipleDefinedSymbolErrors(Context* context, const Scope* scope) {
```

Comment:
What's the reason for wrapping this? Honest question as I've seen it done before but assumed it was generally for wrangling argument types into what the query system allows.

Reply:
I wanted the documented signature to return `void` rather than `const bool&`. (Even if we wanted to get the bool out, I'd want the documented signature to return `bool` rather than `const bool&`. This return type issue is usually the reason I wrap these but occasionally the wrapper does some additional checking before launching the query).